The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/4646393006/jobs/8222787067) [(plain)](https://github.com/rust-lang/rust/commit/c9bb9a5e7a39af70abb0a8f89914b224ab5b95e5/checks/12609761082/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
    Checking std_detect v0.1.5 (/checkout/library/stdarch/crates/std_detect)
    Checking hashbrown v0.12.3
    Checking object v0.29.0
    Checking addr2line v0.17.0
error[E0599]: no method named `snooze` found for struct `Backoff` in the current scope
   --> library/std/src/sync/mpmc/list.rs:559:25
    |
559 |                 backoff.snooze();
    |                         ^^^^^^ method not found in `Backoff`
    |
   ::: library/std/src/sync/mpmc/utils.rs:96:1
96  | pub struct Backoff {
96  | pub struct Backoff {
    | ------------------ method `snooze` not found for this struct
For more information about this error, try `rustc --explain E0599`.
error: could not compile `std` due to previous error
Build completed unsuccessfully in 0:00:21
```

</details>

--- a/library/std/src/sync/mpmc/list.rs
+++ b/library/std/src/sync/mpmc/list.rs
@@ -549,6 +549,18 @@ impl<T> Channel<T> {
         let mut head = self.head.index.load(Ordering::Acquire);
         let mut block = self.head.block.load(Ordering::Acquire);
 
+        // If we're going to be dropping messages we need to synchronize with initialization
+        if head >> SHIFT != tail >> SHIFT {
+            // The block can be null here only if a sender is in the process of initializing the
+            // channel while another sender managed to send a message by inserting it into the
+            // semi-initialized channel and advanced the tail.
+            // In that case, just wait until it gets initialized.
+            while block.is_null() {
+                backoff.spin_heavy();
+                block = self.head.block.load(Ordering::Acquire);
+            }
+        }
+
         unsafe {
             // Drop all messages between head and tail and deallocate the heap-allocated blocks.
             while head >> SHIFT != tail >> SHIFT {

